learn/appendix_opengl/debug_callback.md

```diff
@@ -0,0 +1,158 @@
+"Specification document for AMD_debug_output.") providing this functionality.
+This extensions was later approved by the OpenGL Architecture Review Board (ARB)
+as the extension [`ARB_debug_output`](https://registry.khronos.org/OpenGL/extensions/ARB/ARB_debug_output.txt
+"Specification document for ARB_debug_output.") in 2010\. Finally, since OpenGL
+4.3, the extension is a core part of the API, as well as the extension
+[`KHR_debug`](https://registry.khronos.org/OpenGL/extensions/KHR/KHR_debug.txt
+"Specification document for KHR_debug."). This article covers usage of
+both `ARB_debug_output` extension and the 4.3+ API. `KHR_debug` however covers
+so much more than what was introduced in the original extension.
+
+Registering a Debug Callback
+----------------------------
+
+Firstly, you should create a function which will execute when OpenGL has
+debugging information to relay to your program.
+
+```cs
+using System;
+using System.Runtime.InteropServices;
+using OpenTK.Graphics.OpenGL4; // Or alternatively OpenTK.Graphics.OpenGL
+
+// ...
+
+private static void OnDebugMessage(
+    DebugSource source,     // Source of the debugging message.
+    DebugType type,         // Type of the debugging message.
+    int id,                 // ID associated with the message.
+    DebugSeverity severity, // Severity of the message.
+    int length,             // Length of the string in pMessage.
+    IntPtr pMessage,        // Pointer to message string.
+    IntPtr pUserParam)      // The pointer you gave to OpenGL, explained later.
+{
+    // In order to access the string pointed to by pMessage, you can use Marshal
+    // class to copy its contents to a C# string without unsafe code. You can
+    // also use the new function Marshal.PtrToStringUTF8 since .NET Core 1.1.
+    string message = Marshal.PtrToStringAnsi(pMessage, length);
+
+    // The rest of the function is up to you to implement, however a debug output
+    // is always useful.
+    Console.WriteLine("[{0} source={1} type={2} id={3}] {4}", severity, source, type, id, message);
+
+    // Potentially, you may want to throw from the function for certain severity
+    // messages.
+    if (type == DebugType.DebugTypeError)
+    {
+        throw new Exception(message);
+    }
+}
+```
+
+> [!NOTE]
+> Neither the access modifier or whether the method is an instance method or a
+> static method matters.
+
+> [!WARNING]
+> This function is called from native code, which means there are stack frames
+> associated with them. If your debugger does not support mixed-mode debugging
+> it may affect your user experience if an exception is thrown or occurs within
+> the callback.
+
+> [!CAUTION]
+> The pointer `pMessage` is a a reference to memory you do not own. Attempts to
+> write to or accessing it outside the scope of the function can and will cause
+> an access violation. However, it is safe to use the string copied by the
+> `Marshal` class.
```

Comment:
I don't really think this warning is necessary. Most C# developers will not want to use the raw `IntPtr` and will only ever call `Marshal.PtrToStringAnsi`. To them this warning is only extra noise and terms to understand.